### DISCUSSION TITLE
Question: Using handler decorators on methods

### DISCUSSION THREAD

**Discussion by byte-bot-app** [litestar-org/litestar#3173] [Q&A] [answered]
Reported by @skewty  in Discord: <#1212105750608871445>:

The following MCVE ish code works, but for reasons outlined in [this (and surrounding messages)](https://discord.com/channels/919193495116337154/1212105750608871445/1212127983272202240) OP prefers to use a web socket handler that is a method.

Changing `async def handler(socket: WebSocket) -> None:` to `async def handler(self, socket: WebSocket) -> None:` causes a 403 connection closed error.

```python
from litestar import Litestar, WebSocket, websocket

class Something:
    @websocket("/ws")
    async def handler(socket: WebSocket) -> None:
        await socket.accept()

app = Litestar([Something().handler])
```

As per [docs](https://docs.litestar.dev/latest/usage/routing/handlers.html#websocket-route-handlers)

> In difference to HTTP routes handlers, websocket handlers have the following requirements:

> they must declare a socket kwarg.
> they must have a return annotation of None.
> they must be async functions.

> These requirements are enforced using inspection, and if any of them is unfulfilled an informative exception will be raised.

Shouldn't an error be raised? Is this usage recommend?

**Comment by skewty** [comment]
So I tried to implement this as a stand-alone function and discovered that any dependency named "self" breaks the route. 
Renaming the dependency / function parameter from "self" "self_" fixes the issue and the web socket route starts working again.

**Comment by skewty** [comment]
Also note the handler copies some code from: https://docs.litestar.dev/2/usage/channels.html#managing-history

But when running it:
```
RuntimeWarning: coroutine 'ChannelsPlugin.subscribe' was never awaited
  async with channels.subscribe(all_channels) as subscriber:
```

It appears the sample code in the documentation is no longer valid.

```python
async with channels.subscribe(["some_channel"]) as subscriber:
# should be changed to
async with channels.start_subscription(["some_channel"]) as subscriber:
```

**Comment by provinzkraut** [comment, marked as answer]
The issue here is that at the time the `@websocket` decorator is passed `Something.handler`, the function is still unbound (i.e. the `self` argument won't receive the instance you created automatically, because it doesn't exist yet).

You can fix this by changing how you pass the handlers:

```python
from litestar import Litestar, WebSocket, websocket

class Something:
    async def handler(socket: WebSocket) -> None:
        await socket.accept()

app = Litestar([websocket("/ws")(Something().handler)])
```

Litestar itself does something similar in the `ChannelsPlugin`:

https://github.com/litestar-org/litestar/blob/1332983e09e13571c9a7490540a21f5cf8121f77/litestar/channels/plugin.py#L117-L126

<hr>

> Is this usage recommend?

No, I wouldn't recommend it, and I would advise you to change how you've set up your code that this is a requirement. Usually there are ways to work around this.

**Comment by aranvir** [comment]
You can also make the class a `Controller`, then `self` is not a problem. 

```python
from litestar import Litestar, WebSocket, websocket, Controller

class Something(Controller):
    @websocket("/ws")
    async def handler(self, socket: WebSocket) -> None:
        await socket.accept()

app = Litestar([Something])
```
